src/components/burger-constructor/burger-constructor.tsx

```diff
@@ -1,44 +1,116 @@
+import React, { useContext, useEffect, useReducer, useState } from "react";
 import {ConstructorElement, DragIcon, Button, CurrencyIcon}  from '@ya.praktikum/react-developer-burger-ui-components';
 import styles from './burger-constructor.module.css';
 import Modal from '../modal/modal';
 import OrderDetails from './../order-details/order-details';
 import PropTypes from 'prop-types';
 import ingredientItemPropTypes from '../../utils/ingredient-item-prop-types';
+import IngridientsDataContext from "../services/IngridientsDataContext";
+import { Ingredient } from "../model/ingredient";
+import { cartData } from "../services/cartContext";
+import ConstructorDataContext from "../services/constructorDataContext";
+
+const sentOrderUrl = 'https://norma.nomoreparties.space/api/orders';
 
 function BurgerConstructor(props: any){
 
-  const [data, setData] = useState([] as any[]);
+  const ingredientData = useContext(IngridientsDataContext) as Ingredient[];
   const [orderOpen, setOrderOpen] = useState(false);
 
-  useEffect(() => {
-    setData(props.data);
-  }, [props.data, orderOpen])
+  const [orderData, setOrderData] = useState(null);
+  const [cartState, dispatchCart] = useContext(ConstructorDataContext);
 
+  // Открыть окно заказа
   const openOrder = () => {
-    setOrderOpen(true);
+    let ingredients: any[] = [];
+    cartState.items.forEach((item: Ingredient) => {
+      ingredients.push(item._id);
+    });
+    
+    const options = {
+      method: 'POST',
+      headers: {
+      'Content-Type': 'application/json'
+     },
+      body: JSON.stringify({"ingredients": ingredients})
+    };
+
+    // todo: добавить http класс для отправки и обработки запросов
+    fetch(sentOrderUrl, options as any)
+      .then(res => {
+        if (!res.ok) {
+          throw new Error('ошибка ' + res.status);
+        }
+        return res.json();
+      })
+      .then((json: any) =>   {
+        if (!json.success) {
+          throw new Error('ошибка');
+        }
+        console.log('json', json);
+        setOrderData({
+          name: json.name,
+          number: json.order.number
+        } as any)
+        setOrderOpen(true);
+      })
+      .catch((error) => {
+        console.error('Что то пошло не так... :(', error);
+      })
+
   }
   const closeOrder = () => {
     setOrderOpen(false);
   }
 
+  const initCart = (ingredients: Ingredient[]): void => {
+    let oneBunFinded = false;
+    ingredients = ingredients.filter(item => {
+      if (item.type !== 'bun' || !oneBunFinded) {
+        if (item.type === 'bun') {
+          oneBunFinded = true;
+        }
+        return true;
+      }
+      return false;
+    });
+    dispatchCart(ingredients);
+  }
+  const clearCart = (): void => {
+    dispatchCart([]);
+  }
+  const addItemToCart = (ingredient: Ingredient): void => {
+    // todo: проверить что бы булка была только одна
+    const items = [...cartState.items];
+    items.push(ingredient);
```

Comment:
Тут можно обойтись без метода `push`, так как при создании нового массива на основе другого можно и добавлять новые элементы
```suggestion
    const data = [...oldData, dataItem];
```